simple-proofs/list-algorithms.md

```diff
@@ -0,0 +1,145 @@
+```k
+requires "verification.k"
+
+module LIST-ALGORITHMS
+  imports VERIFICATION
+
+  rule { M:Map [K <- V1] #Equals M [K <- V2] } => { V1 #Equals V2 } [simplification, anywhere]
+
+  rule 
+  {
+    M1:Map [ K1 <- V1  ] [ K2 <- V2  ] #Equals
+    M2:Map [ K1 <- V1' ] [ K2 <- V2' ] 
+  } => 
+  { 
+    true #Equals 
+            V1 ==K V1' andBool V2 ==K V2'
+    andBool M1 [ K1 <- undef ] [ K2 <- undef ] ==K M2 [ K1 <- undef ] [ K2 <- undef ]
+  } [simplification, anywhere]
+
+  rule 
+  {
+    M1:Map [ K1 <- V1  ] [ K2 <- V2  ] [ K3 <- V3  ]
+  #Equals
+    M2:Map [ K1 <- V1' ] [ K2 <- V2' ] [ K3 <- V3' ]
+  } => 
+  { 
+    true #Equals 
+            V1 ==K V1' andBool V2 ==K V2' andBool V3 ==K V3'
+    andBool M1 [ K1 <- undef ] [ K2 <- undef ] [ K3 <- undef ] ==K M2 [ K1 <- undef ] [ K2 <- undef ] [ K3 <- undef ]
+  } [simplification, anywhere]
+
+  // --------------------------------------------------------
+
+  // 1. List-Free Algorithm
+  claim 
+    <k>
+      [ Z LIST_FREE ( con list(integer) [ _XS:ConstantList ] ) ] 
+        => 
+      < con list(integer) [ .ConstantList ] > ... 
+    </k>
+    <env> _ => .Map </env>
+
+  // List-Free Invariant
+  claim 
+    <k>
+      LIST_FREE_LOOP_BODY
+        =>
+      < con list(integer) [ .ConstantList ] > ... 
+    </k>
+    <env>
+      RHO:Map 
+        [ f_lstFree <- < lam y_0 [ x_0 x_0 y_0 ] RHO_1 [ x_0 <- < lam x_0 [ f_0 ( lam y_0 [ x_0 x_0 y_0 ] ) ] RHO_1 > ] > ] 
+        [ in_lst <- < con list ( integer ) [ _XS:ConstantList ] > ] => .Map 
+    </env>
+    requires RHO_1 ==K RHO [ f_0 <- < lam f_lstFree LIST_FREE_BODY RHO > ]
+
+  // --------------------------------------------------------
+
+  // 2. List-Sum Algorithm
+  claim <k>
+    [ Z LIST_SUM ( con list(integer) [ XS:ConstantList ] ) ] 
+      => 
+    < con integer sum(XS) >
+    ... </k>
+    <env> _ => .Map </env>
+    requires allInts(XS)
+
+  // List-Sum Invariant
+  claim 
+    <k>
+      LIST_SUM_LOOP_BODY
+        =>
+      < con integer sum(XS) > ... 
+    </k>
+    <env>
+      RHO:Map 
+        [ f_lstSum <- < lam y_0 [ x_0 x_0 y_0 ] RHO_1 [ x_0 <- < lam x_0 [ f_0 ( lam y_0 [ x_0 x_0 y_0 ] ) ] RHO_1 > ] > ] 
+        [ in_lst <- < con list ( integer ) [ XS:ConstantList ] > ] => .Map 
+    </env>
+    requires RHO_1 ==K RHO [ f_0 <- < lam f_lstSum LIST_SUM_BODY RHO > ]
+     andBool allInts(XS)
```

Comment:
This proof ends on a strange branch at node 74 where that node itself is considered complete and then it's one more step on the other branch to reduce to the `sum(XS)` term on the RHS.

The K cell on node 74:
```
castToInt ( C:Constant ) ~> castToInt ( sum ( _L:ConstantList ) ) ~> < con integer project:Int ( C:Constant ~> . ) +Int sum ( _L:ConstantList ) > ~> _DotVar1:K
```

Reply:
This could be a question for @virgil-serbanuta, because he had the idea to work with `castToInt`. I think that this is due to this:

```
  syntax KItem ::= castToInt(Constant)
  rule <k> castToInt(_:Int)      => .       ... </k>
  rule <k> castToInt(C:Constant) => (error) ... </k> ensures notBool isInt(C) [owise]
```
where the branching always happens, then the first branch is the one that reduces to `sum(XS)` on the RHS, and the second one has an `ensures` that is in contradiction with `{ true #Equals isInt ( C:Constant ~> . ) }`, and so the branch is considered proven. Does this make sense?

Reply:
I think so. It's considered proven because it resolves to `#Bottom`?